simulacrum/dataset.py

```diff
@@ -1,185 +1,97 @@
-            logging.error('cols and types must both be defined together, as lists')
-
-        if (cols and types):
-            validate_types(types)
-            if len(cols) != len(types):
-                logging.error('cols and types must be lists of equal length')
-            for i in len(cols):
-                series_res[col[i]] = ops[types[i]['type']](types[i], length)
-
-        else:
-            if not coltypes:
-                logging.error('please define either cols and types or coltypes')
-            # Assure iteritems compatibility throught 2.7 and 3+
-            try:
-                coltypes_items = coltypes.iteritems()
-            except AttributeError:
-                coltypes_items = coltypes.items()
-            for col, typ in coltypes_items:
-                data_builder = ops[typ['type']]
-                del typ['type']
-                series_res[col] = data_builder(typ, length)
+import logging
+import pandas as pd
 
-        return pd.DataFrame(series_res)
+from simulacrum import types
+
+TYPE_FUNCTIONS = {
+    'num': types.num_data,
+    'int': types.num_int,
+    'norm': types.norm_data,
+    'exp': types.exp_data,
+    'bin': types.binom_data,
+    'pois': types.poisson_data,
+    'txt': types.text_data,
+    'name': types.name_data,
+    'addr': types.address_data,
+    'date': types.date_data,
+    'coords': types.coords_data,
+    'uuid': types.uuid_data,
+    'faker': types.faker_data}
+
+
+def create(length=100, cols=None, types=None, coltypes=None):
+    """Create a dataset based on passed in information.
+
+    A user must either pass in cols and types lists, OR coltypes, OR the
+    function will default to "one of each" type.
+
+    Parameters
+    ----------
+    length : int, optional
+        How many records (rows) in the returned dataframe
+    cols : list, optional
+        A list of column names
+    types : list of dict, optional
+        A list of "type dictionaries", defining both the type, and any
+        parameters to be passed to the type function.
+    coltypes : dict, optional
+        A combined version of cols and types, where the keys of the dictionary
+        are cols, and the values are type dictionaries.
+    """
+    if cols and types and coltypes:
+        raise ValueError(
+            'coltypes should not be defined when cols and types are defined')
+    elif cols is not None and types is not None:
+        column_iter = cols
+        type_iter = types
+        if len(column_iter) != len(type_iter):
+            raise ValueError(
+                'cols and types must be lists of equal length')
+    else:
+        if coltypes is None:
+            logging.warning('No coltypes specified, using default')
+            coltypes = default_coltypes()
+        column_iter = coltypes.keys()
+        type_iter = coltypes.values()
+
+    series_res = {}
+    for col, type_dict in zip(column_iter, type_iter):
+        validate_type_dict(type_dict)
```

Comment:
Might actually be able to just leverage the base `TypeError` that would be raised:

```python
>>> import simulacrum as sm
>>>
>>> test_good = {'entries': {'type': 'exp', 'lam': 0.5}}
>>> res_good = sm.create(10, coltypes=test_good)
>>>
>>> test_bad = {'entries': {'type': 'exp', 'bad_key': 100}}
>>> res_bad = sm.create(10, coltypes=test_bad)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ian/Documents/github/simulacrum/simulacrum/dataset.py", line 64, in create
    series_res[col] = data_builder(length, **type_dict)
TypeError: exp_data() got an unexpected keyword argument 'bad_key'
>>>
>>> test_bad_type = {'entries': {'type': 'norm', 'mean': '0', 'sd': 1}}
>>> res_bad = sm.create(10, coltypes=test_bad_type)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ian/Documents/github/simulacrum/simulacrum/dataset.py", line 64, in create
    series_res[col] = data_builder(length, **type_dict)
  File "/Users/ian/Documents/github/simulacrum/simulacrum/types.py", line 35, in norm_data
    return pd.Series(np.random.normal(mean, sd, length))
  File "mtrand.pyx", line 1643, in mtrand.RandomState.normal (numpy/random/mtrand/mtrand.c:20599)
TypeError: must be real number, not str
```

I think those are decently good error messages to raise to a user, at least the former.  For the latter, if you wanted a better message, you could always raise a specific message in `norm_data` like:

```python
def norm_data(length, mean=0, sd=1):
    """Normal distribution data"""
    try:
        return pd.Series(np.random.normal(mean, sd, length))
    except TypeError:
        raise TypeError('mean ({}) and sd ({}) must both be numeric!'.format(mean, sd))
```

Thoughts?  I do agree that any handling that you want to add could be in a separate PR.